### PR TITLE
Ssprove backend base

### PR DESCRIPTION
**Pull request by cmester0**
The SSProve backend, waiting on #348, ~~#357~~ ~~and #362~~

### CODE REVIEW

**Comment by W95Psp** [comment]
Looks like you need to `cargo test --test toolchain` and then  `cargo insta review` :)

**Comment by cmester0** [comment]
> Looks like you need to `cargo test --test toolchain` and then `cargo insta review` :)

Seems I am getting different results when running the tests locally (instead of with nix). However I think the issue is that I removed reordering for Coq and SSProve, as they put some functions in an incorrect order, while the default order works.

**Comment by cmester0** [comment]
This should be ready for merging. @W95Psp

File: engine/lib/phases/phase_state_passing_loop.ml [hunk 1]
```diff
@@ -0,0 +1,378 @@
+(* TODO: handle Exn report *)
```
Comment (W95Psp) [inline engine/lib/phases/phase_state_passing_loop.ml:1]: Mmh, this is basically a copy-paste of `phase_local_mutation` with a few differences, right?
We need to either make `phase_local_mutation` parametric or split it in two.

I recall we discussed that phase at some point.
If I'm not mistaking, you want to allow local mutation, but at the boundaries of functions and loops? i.e. you want to rewrite mutations of outer variables within a loop?

Currently, I think this `phase_state_passing_loop` is still rewriting all local mutation into shadowing bindings though! Or I'm mistaking?

File: engine/bin/lib.ml [hunk 2]
```diff
@@ -43,7 +43,13 @@ let import_thir_items (include_clauses : Types.inclusion_clause list)
...
       items
   in
-  let items = Deps.sort items in
+  (* TODO: should keep order for non-dependent functions by default? This breaks ssprove extraction. Alternatively this should be a phase *)
+  let items =
+    Deps.sort items
+    (* match options.backend.backend with *)
+    (* | Coq | Ssprove -> items *)
+    (* | _ -> Deps.sort items *)
+  in
```
[the remark this answers is not in the document]
Comment (cmester0) [reply engine/bin/lib.ml:52]: Hmm, seems updates fixed the ordering issue for the generated SSProve code. Should I file an issue about making this into a phase?

File: engine/lib/phases/phase_state_passing_loop.ml (hunk 1 above)
Comment (cmester0) [reply engine/lib/phases/phase_state_passing_loop.ml:1]: Yes, I did copy paste this from `phases_local_mutation`. I want to keep mutable variables (to some extent), the only changes I did was keeping the mutability, as I do not see why it was removed by this phase, however I could refactor it into two phases instead, and just keep the first for the SSProve backend.

Comment (cmester0) [reply engine/lib/phases/phase_state_passing_loop.ml:1]: I might want to allow more mutability, as SSProve can actually reason about it. However, for now I just want to keep the mutable assignments throughout the translations, this is still translated to "shadowing" mutable let binding, but this is an assignment to a memory location by name, so it has the same effect.

**Comment by cmester0** [comment]
@W95Psp we should get this in, we can always do follow up PRs for testing and generalization.

**Review by W95Psp — APPROVED** [review]
OK, let's merge then, it has been waiting for too long anyway!

**Pull request merged**